### ISSUE TITLE
API HTTP 502: "Incomplete response received from application"

### ISSUE THREAD

**Issue by bnewbold** [datacite/datacite#851]
**Describe the bug**

When iterating over all DOI records using the `page[cursor]` API, I received repeated 502 errors.

I have successfully iterated through millions of records. After an earlier error, I was able to resume iteration using the "next URL" field. However, with this error, I am not able to resume iteration; the request always returns an error even after several retries.

I am using the API to iterate over all records because there is not a public bulk metadata dump available, and this technique was mentioned in #188 

**To Reproduce**

Visit the url <https://api.datacite.org/dois?page%5Bcursor%5D=MTQ3MDg5NjI5ODAwMCwxMC4xNTQ2OC9ETC42TEZSRlk&page%5Bsize%5D=1000>

**Expected behavior**

HTTP 200 response with 1000 metadata records.

**Comment by bnewbold** [comment]
I tried iterating again from the beginning and got the same error at:

<https://api.datacite.org/dois?page%5Bcursor%5D=MTQ3MDg5OTIzNjAwMCwxMC4xMzE0MC9SRy4yLjEuMjE5MC42MTYy&page%5Bsize%5D=1000>

In both cases the iteration failed at around 6,143,000 records, or a bit over a third of the corpus.

The later cursor decodes as base64, and the former almost does; if I drop some characters I get something to decode. The cursor looks like a UNIX timestamp + DOI, comma separated. In both cases the time. Eg:

    base64.b64decode('MTQ3MDg5OTIzNjAwMCwxMC4xMzE0MC9SRy4yLjEuMjE5MC42MTYy')
    => b'1470899236000,10.13140/RG.2.1.2190.6162'

The other timestamp was `1470896298000`. Both in August, 2016.

Presumably each pagination is "records registered or updated after this timestamp; first record should match this DOI" or something like that. I tried fudging the UNIX timestamp to jump ahead by a day or week to see if there was just a single bad DOI registered that was causing problems, but this didn't seem to work (maybe because the DOI part didn't match).

Regardless, there seems to be no point trying again as the same issue will be encountered. Both the sitemap.xml and API iteration methods seem broken, so i'm at a loss at how to get a list of all Datacite DOIs, or better, a dump of metadata.

**Comment by mfenner** [comment]
@bnewbold I am sorry for the trouble you have with this. This is unfortunately a known problem that is not so easy to fix. The issue is cursor-based paging in Elasticsearch where at some point the paging breaks, most likely because Elasticsearch runs out of memory. The fix for now is to break your query into smaller chunks (e.g. "all DOIs for 2012" and/or "all dois of resourceTypeGeneral dataset") using query parameters.

In a few weeks I have set aside time to properly fix this, and to generate a data dump for all DartaCite DOIs that can be downloaded. The latter should be done before the end of the year.

**Comment by richardhallett** [comment]
Closing in favour of our main issue: https://github.com/datacite/lupo/issues/237 - Please track this ticket for resolution.

Also as Martin mentioned you can attempt to narrow down your querying docs:
https://support.datacite.org/docs/api-queries
Example using updated field:
https://api.datacite.org/dois?query=updated:[2019-01-01T00:00:00Z+TO+2019-02-01T23:59:59Z]&state=findable

Many thanks again for reporting.

**Comment by miku** [comment]
Just a short addendum: With smaller (24h) slices it is still possible to occasionally run into a HTTP 502 - [example](https://git.io/JeKWv).

```
[2019-02-21T00:00:00Z TO 2019-02-21T23:59:59Z]
```

**Comment by mfenner** [comment]
This can also be an issue with smaller slices when there is a lot of metadata. The underlying seems to be problem is that processing takes too long, at some point bringing down the server.
